core/src/main/java/net/adoptopenjdk/icedteaweb/resources/downloader/PrioritizedParallelExecutor.java

```diff
@@ -0,0 +1,220 @@
+     */
+    public PrioritizedParallelExecutor(ExecutorService executor) {
+        this.executor = executor;
+    }
+
+    /**
+     * Executes the passed callables in parallel
+     * The callables must be sorted by priority with the highest priority as the first in list.
+     * Executions of lower priority are canceled as soon as a higher priority result has successfully been calculated.
+     * The returned {@link Future} will complete as soon as the first callable completed successfully and all other
+     * callables with higher priority (if any) have completed exceptionally.
+     *
+     * @param callables the callables to execute.
+     * @return a future holding the best result from all callables.
+     */
+    public <V> Future<V> getSuccessfulResultWithHighestPriority(List<Callable<V>> callables) {
+        if (callables == null || callables.isEmpty()) {
+            throw new IllegalArgumentException("No callables");
+        }
+
+        if (callables.size() == 1) {
+            return executor.submit(callables.get(0));
+        }
+
+        final CancelingCallable<V> head = new CancelingCallable<>(callables.get(0));
+        final List<Callable<V>> tail = callables.subList(1, callables.size());
+
+        final ChainableFuture<V> result = new ChainableFuture<>(executor.submit(head));
+        final Future<V> lowerPriority = getSuccessfulResultWithHighestPriority(tail);
+
+        head.setLowerPriority(lowerPriority);
+        result.setLowerPriority(lowerPriority);
+
+        return result;
+    }
+
+    /**
+     * Wrapper around {@link Callable} which will cancel any lower priority after completing the call successfully.
+     */
+    private static class CancelingCallable<V> implements Callable<V> {
+
+        private final AtomicBoolean complete = new AtomicBoolean(false);
+        private final AtomicReference<Future<V>> lowerPriority = new AtomicReference<>();
+
+        private final Callable<V> delegate;
+
+        private CancelingCallable(Callable<V> delegate) {
+            this.delegate = delegate;
+        }
+
+        /**
+         * Set a lower priority to be canceled.
+         * This is not passed in the constructor as the value is not known then.
+         * <p>
+         * If this callable has already been completed successfully then the passed future will be canceled immediately.
+         *
+         * @param lowerPriority the lower priority future to be canceled after completing this callable successfully.
+         */
+        void setLowerPriority(Future<V> lowerPriority) {
+            this.lowerPriority.set(lowerPriority);
+            if (complete.get()) {
+                cancelLowerPriority();
+            }
+        }
+
+        @Override
+        public V call() throws Exception {
+            final V result = delegate.call();
+            complete.set(true);
+            cancelLowerPriority();
+            return result;
+        }
+
+        private void cancelLowerPriority() {
```

Comment:
Since this is called in the `call()` method and in the `setLowerPriority(...)` method it could be called twice (even parallel) based on a race condition.  Based on this the lowerPriority can be canceled more than once. I do not know if this might end in problems.

Reply:
cancel will simply return false if the future is already canceled